### PR TITLE
[stdlib] Introduce ContiguousCollection protocol

### PR DESCRIPTION
**Pull request by airspeedswift**
A protocol to unify various contiguously-stored collections that offer with-unsafe-buffer capabilities.

### CODE REVIEW

**Comment by airspeedswift** [comment]
@swift-ci please benchmark

File: stdlib/public/core/ContiguousCollection.swift [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    let i = _base.distance(from: _base.startIndex, to: startIndex)
+    let j = _base.distance(from: _base.startIndex, to: endIndex)
+    return try _base.withUnsafeMutableBufferPointer {
+      var buffer = UnsafeMutableBufferPointer(rebasing: $0[i..<j])
+      return try body(&buffer)
+    }
+  }
+}
+
+extension MutableContiguousCollection where Self: MutableContiguousCollection {
```
Comment (harlanhaskins) [inline stdlib/public/core/ContiguousCollection.swift]: Nitpick: Is this a redundant constraint?

Comment (airspeedswift) [reply stdlib/public/core/ContiguousCollection.swift]: oops, yes

Comment (harlanhaskins) [reply stdlib/public/core/ContiguousCollection.swift]: Seems like it's worth a warning.  https://bugs.swift.org/browse/SR-9218

**Comment by airspeedswift** [comment]
@swift-ci please benchmark

**Comment by airspeedswift** [comment]
@swift-ci please benchmark

**Comment by airspeedswift** [comment]
@swift-ci please benchmark

**Comment by airspeedswift** [comment]
@swift-ci please benchmark

**Pull request closed without merging**